src/shared/group-record.c

```diff
@@ -347,6 +347,9 @@ bool group_record_matches_group_name(const GroupRecord *g, const char *group_nam
 int group_record_match(GroupRecord *h, const UserDBMatch *match) {
         assert(h);
 
+        if (!gid_is_valid(h->gid))
+                return -EINVAL;
+
```

Comment:
hmm, i think we should not consider this an error (it's valid for group records to have no gid, if they aren't "fixated" locally yet), but simply as a case of non-matching. something that has no gid, doesn't have a gid between two boundary values either, hence is not matching.

Hence, I'd prefer if we'd move this down, after the `!match` check and simply turn into:

```c
if (!gid_is_valid(h->gid)) /* a user record without gid, definitely doesn't have a gid between two ranges */
        return false;
```
```

Reply:
That was the previous version, and I proposed the current form above...

But at least there's an inherent inconsistency in the existing impl: user/group_record_match returns a int, and we do check for error at various places. Either it should be made clear that they're not allowed to fail, or handle them properly.

Plus, I don't see any actual code path where a record without valid uid/gid might be accepted by groupdb_iterator_get_one. Can you clarify?

Reply:
So, I returned to the original variant and modified callers instead since there's never an error from the matching functions.
Please provide 2nd pair of eyes, since callers are almost same but not all of them.

Reply:
> But at least there's an inherent inconsistency in the existing impl: user/group_record_match returns a int, and we do check for error at various places. Either it should be made clear that they're not allowed to fail, or handle them properly.

they should be allowed to fail with things like ENOMEM only really